### PR TITLE
Fix span sampling

### PR DESCRIPTION
**Pull request by Qard**
Fix several bugs in the single span ingestion implementation

Bugs fixed:
- `SpanSampler` was not receiving `config.sampler`
- `SpanContext` sampling object was being reference-copied which made descending span contexts mutate their parents
- `Limiter` should should only be used it the `max_per_second` input is present and a valid finite number
- The `Limiter` should be associated with the matched rule, not with the `service:name` pair of the input context
- The `service` to match against should be pulled from span tags, if present
- An absent `service` or `name` on the rule should auto-match without running the glob matcher

### CODE REVIEW

**Review by uurien — COMMENTED** [review]
The changes looks good for me, but I think it is worth adding some tests

**Review by juan-fernandez — COMMENTED** [review]
+1 to @uurien. Any chance to add tests to this?

**Comment by Qard** [comment]
It's covered by system-tests, but I can see about adding some unit tests too. There's already a bunch of _other_ unit tests and it still passes those, but that probably also just demonstrates the insufficiency of the prior unit tests. 😅

**Comment by Qard** [comment]
Added tests for each of the fixed bugs and also updated a bunch of the existing tests to assert non-matching where relevant.

**Review by uurien — APPROVED** [review]

**Comment by Qard** [comment]
I will wait for DataDog/system-tests#848 to land and then rebase DataDog/system-tests#858 before landing this just to be sure any changes to the system tests are still passing.

**Review by rochdev — APPROVED** [review]

**Pull request merged**
